### PR TITLE
Minimal server search / enable improved focus mode

### PR DESCRIPTION
**Pull request by marcaaron**
### Details

Adds server assisted searching to the "Search" page.

- [x] Holding on https://github.com/Expensify/Auth/pull/8725
- [x] Holding on https://github.com/Expensify/Web-Expensify/pull/38916 

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/318005
$ https://github.com/Expensify/Expensify/issues/318006

### Tests

The tests here are primarily designed to prove that we can return reports that hit on the following search queries from Auth...

- A participant login (for now, we are not going to be able to search by display name)
- A policy name (it should return any newdot accessible reports on the policy)
- Report owner email (e.g. a workspace chat)

1. Sign into NewDot
2. Create a workspace
3. Rename the workspace to "Business Name"
4. Add 2 employees
6. Navigate to the Search page
7. Enter the search term "concierge"
8. Verify there is a loading spinner that appears inside the text input and then is removed later once the API command finishes
9. Verify via the Chrome DevTools that the Onyx data response for the `SearchForReports` command returns the Concierge report

11. Enter the search term "test"
12. Verify that the 3 workspace chats and 2 policy rooms are returned by `SearchForReports`
13. Enter a specific login for one of the employees added
14. Verify the "#announce" report and the workspace chat that they own are returned in the response

15. Sign in as one of the employees and set the `firstName` field of the personal details to "Bob"
16. Sign back in as the workspace admin
17. Navigate to Search again
18. Enter "Bob" into the search field
19. Verify the same result as step 14.

**Test that switching from "#focus" to "Most recent" will restore chats**

1. Switch to focus mode
2. Sign out
3. Sign in
4. Verify that only a limited number of chats are displayed
5. Set up two other devices and sign in with the same account
6. Leave one device open and the other backgrounded (mobile)
7. Switch to "Most recent" mode on the main account from "focus" mode
8. And verify that...
9. All the devices switched modes and have all the chats

- [ ] Verify that no errors appear in the JS console

### Offline tests

1. Go offline while in the Search view
10. Verify that a message appears indicating that search results will be limited

### QA Steps

1. Same as tests

(probably Internal so that we can help verify)

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by marcaaron** [comment]
@roryabraham @aldo-expensify taking this out of draft to start some reviews...

Still waiting on:

- [ ] Correct español translation
- [ ] Staging CPs for the 2 web PRs to opt into the focus mode and `SearchForReports` API command

It's been testing well for me locally more or less. 

One thing that was missed is that when we switch the priority mode from "#focus" to "Most recent" we need to return all the chats so that view can be populated. This situation is possible:

1. Be signed into two devices with "#focus" mode
2. Put one into the background
3. Update the priority mode to "Most recent" on one device
4. Open the other device up again
5. Other device will be in "Most recent" mode, but still only have the "#focus" mode chats

The obvious solution seems to be to just get and send all the chats via Pusher when the priority mode changes, but I am not actually sure if it would be a problem to send these large payloads via Pusher... will think more about it. I think it's not much of a blocker to do some testing on an Ad Hoc build. But maybe we want to hold merging this until there's a solution for that.

My plan after the review is done is to do an Ad Hoc build - but not merge this PR until:

- We have a few guides test to see if things are improved for them
- Things are stable and tested
- Design kinks related to search are ironed 
- The switch between "#focus" and "Most recent" is a bit more polish

**Comment by marcaaron** [comment]
as we go, I will also add some more manual test steps for checking to see that the correct focus mode chats will come back when signing in.

**Comment by sakluger** [comment]
Looks like https://github.com/Expensify/Web-Expensify/pull/38916 is now on staging 🎉 . Does that mean this PR is now off hold?

File: src/libs/actions/Report.js
Comment (aldo-expensify) [reply src/libs/actions/Report.js:2213]: Agree with this, I think `_.debounce` is the right thing

Comment (marcaaron) [reply src/libs/actions/Report.js:2213]: Cool, I will try this instead.

Comment (marcaaron) [reply src/libs/actions/Report.js:2213]: I think my original reasoning was that we would want to just start calling the API consistently over time instead of waiting some delay. But didn't think that the characters could get dropped.

File: src/pages/NewChatPage.js
Comment (marcaaron) [reply src/pages/NewChatPage.js:175]: I think it sounds great. And more consistent with how I handled it in the Search page.

File: src/pages/SearchPage.js
Comment (marcaaron) [reply src/pages/SearchPage.js:195]: Good catch. Yes, I want them to be the same.

**Comment by aldo-expensify** [comment]
@marcaaron no need to keep holding on https://github.com/Expensify/Web-Expensify/pull/38916, right? since it hit staging already

**Comment by marcaaron** [comment]
Yes it's off hold now. I need to fix a few things here, but will request re-reviews soon. Thanks!

**Comment by marcaaron** [comment]
I left some comments here related to the UX: https://expensify.slack.com/archives/C04JD99283A/p1696373941266199

Would like to discuss them a bit before merging this one.

**Comment by marcaaron** [comment]
Design conversation seems wrapped for now. I like where we landed. Probably can improve things in the future. I re-tested everything on all the platforms.

@aldo-expensify but I think we are still blocked on [the conversation here related to how to handle the switch from "#focus" to "Most recent" mode](https://expensify.slack.com/archives/C04JD99283A/p1696394329948359?thread_ts=1696296312.806639&amp;cid=C04JD99283A).

**Comment by marcaaron** [comment]
Based on the [conversation in Wave1](https://expensify.slack.com/archives/C04JD99283A/p1696537236706269?thread_ts=1696296312.806639&cid=C04JD99283A) I think we can:

- proceed with a naive solution and just call OpenApp when the priority mode changes
- follow up with an optimized command that only will load the reports and call that instead

Lmk if it sounds OK to you @aldo-expensify and if the solution here seems OK for now.

**Comment by marcaaron** [comment]
Updated if we can get some final reviews here 🙇

**Comment by aldo-expensify** [comment]
Reviewing 👀

File: src/pages/NewChatPage.js
Comment (aldo-expensify) [inline src/pages/NewChatPage.js:178]: NAB: Wondering if we should have kept `onChangeText={setSearchTerm}` and use a `useEffect`

```
useEffect(() => {
        if (searchTerm.length) {
            Report.searchInServer(searchTerm);
        }
}, [searchTerm]);
```
To do trigger the server search. I think I like the `useEffect` pattern more, but NAB!

**Review by aldo-expensify — DISMISSED** [review]
Works well for me!

File: src/pages/NewChatPage.js
Comment (aldo-expensify) [reply src/pages/NewChatPage.js:178]: Hehe, I see this is exactly the opposite of @hoangzinh 's preference: https://github.com/Expensify/App/pull/27819#discussion_r1340994924

Comment (aldo-expensify) [reply src/pages/NewChatPage.js:178]: For me it makes sense to use `useEffect` because I see it as a side effect of the the `searchTerm`  changing. Also a benefit of `useEffect` is that you could cancel the debounced call if the components is unmounted... like:

```js
useEffect(() => {
        if (searchTerm.length) {
            Report.searchInServer(searchTerm);
        }
        return () => Report.searchInServer.cancel(); // https://stackoverflow.com/a/41286377
}, [searchTerm]);
```

but maybe we don't care about canceling

* the cancel example above is not working because you have to call `.cancel` on the debounced function and `Report.searchInServer` is not the debounced function... but it is just to show the idea.

**Comment by marcaaron** [comment]
Thank you for all the great review points. I think we are close so I am going to trigger an Ad Hoc build. 

And someone needs to do the checklist 😄

File: src/pages/NewChatPage.js
Comment (marcaaron) [reply src/pages/NewChatPage.js:178]: Personally, I don't think we need to optimize the case of canceling the debounced function - but maybe would if it was directly calling a `setState()` method after the component has unmounted. 

In this case, the chats will load with or without the page and it won't really have any downsides.

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js]: Great thanks!

File: src/libs/actions/App.js
Comment (marcaaron) [reply src/libs/actions/App.js]: Good call.

**Review by hoangzinh — APPROVED** [review]
LGTM, great work @marcaaron

File: src/pages/NewChatPage.js
Comment (Beamanator) [inline src/pages/NewChatPage.js:48]: Lets add this new prop to the propTypes definition in this file too?

File: src/pages/SearchPage.js
Comment (Beamanator) [inline src/pages/SearchPage.js:204]: should we also add this new prop to propTypes?

**Review by Beamanator — CHANGES_REQUESTED** [review]
Code looks good, but we're missing a few new proptypes, no?

File: src/components/OptionsList/BaseOptionsList.js
Comment (marcaaron) [reply src/components/OptionsList/BaseOptionsList.js:248]: William Strunk Jr over here...

**Comment by marcaaron** [comment]
@Beamanator thanks for the review ❤️ ! Made the changes 🙇

File: src/components/OptionsList/BaseOptionsList.js
Comment (Beamanator) [reply src/components/OptionsList/BaseOptionsList.js]: Woah now, you callin me hoity toity just cuz I often enjoy a nice, crisp, full bodied glass of mug root beer??

**Review by Beamanator — DISMISSED** [review]
Looks great and is working well here! But there's conflicts :'(

And I wonder if @roryabraham wants to review before I merge? Aldo is OOO i beleive

**Comment by marcaaron** [comment]
I think they are both OOO

File: src/components/OptionsList/BaseOptionsList.js
Comment (marcaaron) [reply src/components/OptionsList/BaseOptionsList.js]: Well, sippin' on mug root beer doesn't make you hoity toity, but if you start sippin' it with your pinky out, we might need to reevaluate!

-ChatGPT

File: src/libs/actions/App.js
Comment (roryabraham) [inline src/libs/actions/App.js:51]:
```suggestion
        // When someone switches their priority mode we need to fetch all their chats because only #focus mode works with a subset of a user's chats. This is only possible via the OpenApp command.
```

File: src/libs/actions/Report.js
Comment (roryabraham) [inline src/libs/actions/Report.js:2257]: NAB

```suggestion
function searchForReportsInServer(searchInput) {
```

File: src/pages/SearchPage.js
Comment (roryabraham) [inline src/pages/SearchPage.js:42]: Missing JSDoc for prop

File: src/pages/SearchPage.js
Comment (roryabraham) [inline src/pages/SearchPage.js:239]: I think we should use the `withNetwork` HOC instead of `withOnyx`. Also `ONYXKEYS.NETWORK` should typically be `initWithStoredValues: false`, because we don't have any reason to think that the persisted value for `isOffline` should be accurate the next time the user opens the app.

**Review by roryabraham — COMMENTED** [review]
Not officially requesting changes because I don't want to delay this while I'm OOO, but I had a few comments that maybe should be addressed. Don't think anything will blow up if they're not addressed

File: src/pages/SearchPage.js
Comment (marcaaron) [reply src/pages/SearchPage.js:239]: This is a class component still.

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:2257]: it's already `Report.searchInServer()` so you would have `Report.searchForReportInServer()` 🤷

File: src/pages/SearchPage.js
Comment (marcaaron) [reply src/pages/SearchPage.js:239]: Updating to withNetwork since it will reduce a subscription.

But there are so many places where we do not consider this...

>we don't have any reason to think that the persisted value for isOffline should be accurate the next time the user opens the app

nothing super bad happens AFAICT

**Comment by marcaaron** [comment]
Updated!

File: src/libs/actions/Report.js
Comment (Beamanator) [reply src/libs/actions/Report.js:2257]: true that reads a bit funky

**Review by Beamanator — APPROVED** [review]
LGTM!

**Pull request merged**
